
**Diplomatura en Ciencia de Datos, Aprendizaje Automático y sus Aplicaciones**

**Edición 2023**

---

# Trabajo práctico entregable - Parte 2

In [ ]:
import io
import matplotlib
import matplotlib.pyplot as plt
import numpy
import pandas as pd
import seaborn

seaborn.set_context('talk')

## Lectura del dataset

En la notebook 00 se explican los detalles de la siguiente sección.

In [ ]:
url = 'https://raw.githubusercontent.com/DiploDatos/AnalisisyVisualizacion/master/sysarmy_survey_2022_processed.csv'
df = pd.read_csv(url)

In [ ]:
df[:3]
df.shape

(5358, 44)

In [ ]:
df.profile_gender.unique()

array(['Varón Cis', 'Varón cis', 'Mujer Cis', 'Mujer', 'Varón Trans',
       'Queer', 'No binarie', 'Mujer cis', 'Mujer Trans', 'mujer', 'Gay',
       'Fluido', 'Bigénero', 'Femenino'], dtype=object)

In [ ]:
df.loc[:,'profile_g'] = df.profile_gender.replace({'Varón Cis': 'Varón cis','Mujer': 'Mujer cis','Mujer Cis': 'Mujer cis','Femenino': 'Mujer cis','mujer': 'Mujer cis','Mujer':'Mujer cis','Queer':'Diversidades','Varón Trans':'Diversidades','No binarie':'Diversidades','Mujer Trans':'Diversidades','Fluido':'Diversidades','Bigénero':'Diversidades','Gay':'Diversidades'}).fillna(False)

Aplicamos la misma limpieza de datos que en la parte 1 del trabajo, definiendo el data frame df22

In [ ]:
relevant_columns_2 = ['profile_g' , 'salary_monthly_NETO','work_dedication','salary_monthly_BRUTO','work_years_in_company']
df.loc[:,'profile_g'] = df.profile_gender.replace({'Varón Cis': 'Varón cis','Mujer': 'Mujer cis','Mujer Cis': 'Mujer cis','Femenino': 'Mujer cis',
                                                   'mujer': 'Mujer cis','Mujer':'Mujer cis','Queer':'Diversidades',
                                                   'Varón Trans':'Diversidades','No binarie':'Diversidades','Mujer Trans':'Diversidades',
                                                   'Fluido':'Diversidades','Bigénero':'Diversidades','Gay':'Diversidades'}).fillna(False)

df22 = df[relevant_columns_2]
df22.isna().sum()
df22 = df22.dropna()

df22 = df22[df22['work_dedication'] == 'Full-Time']
df22.value_counts(['work_dedication'])

print('El valor máximo del salario mensual bruto es =', numpy.max(df['salary_monthly_BRUTO']))
print('El valor máximo del salario mensual negto es =', numpy.max(df['salary_monthly_NETO']))

df22 = df22[df22['salary_monthly_BRUTO'] != 28000000]
df22 = df22[df22['salary_monthly_NETO'] != 15000000]

El valor máximo del salario mensual bruto es = 28000000.0
El valor máximo del salario mensual negto es = 15000000.0


In [ ]:
df22['work_years_in_company'] = df22['work_years_in_company'].round()
df22 = df22[df22['work_years_in_company'] != 80]
print(df22['work_years_in_company'].unique())

[ 0.  2.  1. 30.  3.  5. 20. 10. 12.  6. 11.  9. 19.  7. 18.  4. 17. 22.
  8. 14. 24. 15. 16. 13. 23. 21. 25. 26. 40. 28. 34. 29. 27.]


In [ ]:
df.profile_g.unique()

array(['Varón cis', 'Mujer cis', 'Diversidades'], dtype=object)

In [ ]:
df[['profile_g', 'salary_monthly_NETO']].groupby('profile_g').describe()

salary_monthly_NETO                                               \
                           count           mean            std  min       25%   
profile_g                                                                       
Diversidades                98.0  261873.543041  356719.368150  0.0   90500.0   
Mujer cis                  929.0  188903.430667  184912.713803  0.0   91000.0   
Varón cis                 4331.0  296252.343889  433193.328154  0.0  125000.0   

                                              
                   50%       75%         max  
profile_g                                     
Diversidades  174448.5  262250.0   1950000.0  
Mujer cis     160000.0  237003.0   1878500.0  
Varón cis     212146.0  307988.0  15000000.0

In [ ]:
df[df.salary_monthly_NETO > 1000]\
  [['profile_g', 'salary_monthly_NETO']].groupby('profile_g').describe()

salary_monthly_NETO                                        \
                           count           mean            std     min   
profile_g                                                                
Diversidades                85.0  301914.321765  367031.242344  2800.0   
Mujer cis                  806.0  217708.858077  182050.554847  1400.0   
Varón cis                 3975.0  322777.029197  442612.347688  1100.0   

                                                         
                   25%       50%        75%         max  
profile_g                                                
Diversidades  125000.0  190000.0  280000.00   1950000.0  
Mujer cis     116000.0  180000.0  248193.75   1878500.0  
Varón cis     147106.5  225000.0  322827.50  15000000.0

In [ ]:
alpha = 0.05

In [ ]:
from pandas.core import describe
is_man = df.profile_g == 'Varón cis'

groupA = df[(df.salary_monthly_NETO > 1000) & is_man].salary_monthly_NETO
groupB = df[(df.salary_monthly_NETO > 1000) & ~is_man].salary_monthly_NETO

EstadisticasA = groupA.describe ()
print(EstadisticasA)

count    3.975000e+03
mean     3.227770e+05
std      4.426123e+05
min      1.100000e+03
25%      1.471065e+05
50%      2.250000e+05
75%      3.228275e+05
max      1.500000e+07
Name: salary_monthly_NETO, dtype: float64


In [ ]:
EstadisticasB = groupB.describe ()
print(EstadisticasB)

count    8.910000e+02
mean     2.257419e+05
std      2.080963e+05
min      1.400000e+03
25%      1.167500e+05
50%      1.800000e+05
75%      2.500000e+05
max      1.950000e+06
Name: salary_monthly_NETO, dtype: float64


## Ejercicio 1: Estimación

**Consigna:**  Calcular una estimación puntual y un intervalo de confianza de nivel (1-alpha) para la resta entre la media del salario Neto para Hombres menos la media del salario Neto para otros géneros(diferencia de las medias entre el grupoA y grupoB). 
¿Cómo se relaciona este intervalo de confianza con el test de hipótesis?

In [ ]:
media_groupA = groupA.mean()
media_groupB = groupB.mean()
desv_est_group_A = groupA.std()
desv_est_group_B = groupB.std()

#Cantidad de observaciones por grupo 
tamaño_A = len(groupA)
tamaño_B = len(groupB)

#Diferencia de medias de los grupos
dif_medias_AB = media_groupA - media_groupB
print('Diferencia de medias =', int(dif_medias_AB))

#Definimos el nivel de significancia al 5% con dos colas
z_0975 = 1.96

#Límite inferior de confianza
Lim_inf_dif_medias_AB =  dif_medias_AB - z_0975 * ((( desv_est_group_A **2 )/ tamaño_A) + (( desv_est_group_B **2 )/ tamaño_B )) ** 0.5
print('Limite inferior del intervalo de confianza =', int(Lim_inf_dif_medias_AB))

#Límite superior de confianza
Lim_sup_dif_medias_AB =  dif_medias_AB + z_0975 * ((( desv_est_group_A **2 )/ tamaño_A) + (( desv_est_group_B **2 )/ tamaño_B )) ** 0.5
print('Limite superior del intervalo de confianza =', int(Lim_sup_dif_medias_AB))

Diferencia de medias = 97035
Limite inferior del intervalo de confianza = 77643
Limite superior del intervalo de confianza = 116426


La estimación puntual para la resta entre la media del salario neto para hombres y el salario neto para el resto de los grupos es 97035;  en la que al considerar un nivel de confianza del 95% el límite superior será 116.426 y el inferior 77.643.



## Ejercicio 2: Test de hipótesis



### 2.1 Formalización

Describir formalmente los distintos componentes de un test de hipótesis para comprobar si la distribución de los salarios es distinta entre los grupos A y B.

**Hipótesis Nula**

$H_0=...$

**Estadístico (Pivote)**
  * Identificar el estadístico
  * Escribir qué distribución tiene bajo $H_0$


En caso de querer conocer si la distribución de los salarios es distinta entre los grupos A y B, optaremos por comparar sus medias. En tal caso plantearemos como hipótesis nula el hecho que ambas medias sean iguales y como hipótesis alternativas que son distintas. 

El estadístico para tal caso será:

media_groupA - media_groupB 

La distribución que tendrá nuestro estadístico de T de student con 4864 grados de libertad (tamaño del grupo A y B menos 2).


### 2.2 P-valor

1. Calcule el p-valor y decida si rechazar o no la hipótesis nula.
2. Interprete el resultado.

Links útiles:
* [Test de hipótesis usando scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)
* [Test de Welch](http://daniellakens.blogspot.com/2015/01/always-use-welchs-t-test-instead-of.html)

In [ ]:
from scipy.stats import ttest_ind
# test de diferencia de medias
t_stat, p_value = ttest_ind(groupA, groupB)
# resultado del test
print('t-statistic:', t_stat)
print('p-value:', p_value)

t-statistic: 6.387301602375005
p-value: 1.84592290312122e-10


Dado que el p-value es inferior al nivel de significación alpha elegido; decidimos rechazar la hipótesis nula de igualdad de medias, concluyendo que existe evidencia para decir que el salario de los hombres es mayor al resto de los géneros


### [Opcional] 2.3 Potencia del test

Nuestra muestra, ¿era lo suficientemente grande para detectar si existe o no una diferencia entre los grupos?

1. Utilice la función `tt_ind_solve_power` para calcular el tamaño necesario de la muestra para un poder estadístico de 0.8, 0.9 y 0.95, asumiendo una significancia estadística de 0.05.
2. ¿Cómo intepretan el poder estadístico de un test? Dado su conocimiento de dominio sobre los datos, ¿les parece que esta muestra es lo suficientemente grande para ser representativo de la tendencia general? ¿y para utilizarlo en un juicio penal contra una empresa XX por una causa de discriminación?

[Documentación](https://www.statsmodels.org/stable/generated/statsmodels.stats.power.tt_ind_solve_power.html)

NOTA: este análisis debería hacerse ANTES de recolectar los datos.

In [ ]:
from statsmodels.stats.power import tt_ind_solve_power

In [ ]:
effect_size = (groupA.mean() - groupB.mean()) / groupB.std()
# nobs1=None  - What we want to know
alpha = 0.05
ratio = len(groupB) / len(groupA)
power= 0.8

In [ ]:
for i in [0.8, 0.9, 0.95]:
  tt = tt_ind_solve_power(effect_size=effect_size, alpha=alpha,power=i, ratio=ratio)
  print('Para poder estadistico =', i, 'el numero de observacioes es igual a', int(tt))

Para poder estadistico = 0.8 el numero de observacioes es igual a 198
Para poder estadistico = 0.9 el numero de observacioes es igual a 265
Para poder estadistico = 0.95 el numero de observacioes es igual a 327


In [38]:
print('El tamaño de A es', tamaño_A)
print('El tamalo de B es', tamaño_B)

El tamaño de A es 3975
El tamalo de B es 891


Observamos que el tamaño de muestra necesario para poder estadistico de 0.8, 0.9 y 0.95 es menor al tamaño de A y de B. La muestra es lo suficientemente grande para detectar si hay diferencia entre los grupos.

## Ejercicio 3: Comunicación y visualización

**Consigna:** Seleccionen un resultado que les parezca relevante a partir de alguno de los ejercicios del entregable. Diseñe e implemente una comunicación en base a este mensaje, en un archivo PDF.

Elija las palabras y visualización más adecuada para que la comunicación sea entendible, efectiva y se dapte a UNA de las siguientes situaciones:

1. Una sección en un artículo de difusión a presentar como parte de una organización sin fines de lucro.
No más de 1 página A4 (o dos si los gráficos son muy grandes).
  1. Ejemplo: Alguna de las secciones [Los ecosistemas de emprendimiento de América Latina y el Caribe frente al COVID-19: Impactos, necesidades y recomendaciones](https://publications.iadb.org/es/los-ecosistemas-de-emprendimiento-de-america-latina-y-el-caribe-frente-al-covid-19-impactos-necesidades-y-recomendaciones), por ejemplo la sección *2.2. Reacciones de los emprendedores*.
  2. Ejemplo: Alguna de las secciones de [The state of gender pay gap in 2021](https://www.payscale.com/data/gender-pay-gap?tk=carousel-ps-rc-job)
  3. Puntos clave: 
    1. Simpleza de los gráficos.
    2. Comunicación en lenguaje simple a personas que no son necesariamente expertos de dominio.
    2. Selección de UNA oración sobre la que se hace énfasis.
    3. No es necesario que mencionen objetivos ni descripciones del conjunto de datos, se supone que eso ya estaría explicado en otras secciones del informe.

2. Una publicación científica o reporte técnico interno. No más de una página A4:
  2. Ejemplo: La sección de resultados de [IZA DP No. 12914: The Impact of a Minimum Wage Change on the Distribution of Wages and Household Income](https://www.iza.org/publications/dp/12914/the-impact-of-a-minimum-wage-change-on-the-distribution-of-wages-and-household-income).
  2. Ejemplo: Alguna de las secciones de [Temporary reduction in daily global CO2 emissions during the COVID-19 forced confinement](https://www.nature.com/articles/s41558-020-0797-x)
  3. Puntos clave:
    3. Nivel de detalle técnico requerido. Es necesario justificar la validez del análisis.
    4. La idea presentada puede ser más compleja. Pueden asumir que la audiencia tiene conocimiento técnico y va a analizar las visualizaciones en detalle. 
    5. Pueden presentar más en detalle las limitaciones del análisis (significancia estadística, etc.)
    2. No es necesario que mencionen objetivos ni descripciones del conjunto de datos, se supone que eso ya estaría explicado en otras secciones del informe.

3. Un tweet (o post de LinkedIn) para la cuenta de su empresa consultora que hace análisis de datos. El objetivo es promocionar un análisis de datos abiertos que van a incluir en su portfolio:
  1. Ejemplo: [Comparación vacunas covid](https://twitter.com/infobeautiful/status/1381577746527236098?s=20)
  2. Ejemplo: [Tweet del BID](https://twitter.com/el_BID/status/1388508583944507396?s=20). Lo valioso de este tweet es que usaron un único número para transmitir un mensaje. Puede ser algo así, o con un gráfico muy simple.
  3. Ejemplo: [Cambio climático](https://twitter.com/UNFCCC/status/1387732156190011394?s=20) Es un muy buen ejemplo, excepto que el gráfico no se lee nada y hay que entrar a la publicación original.
  3. Ejemplo: [¿Cuánto están los programadores en las empresas?](https://www.linkedin.com/posts/denis-rothman-0b034043_tech-career-work-activity-6793861923269054464-gS6y) (No verificamos la veracidad o seriedad de la fuente).
  4. Puntos clave:
    1. Su audiencia no va a mirar la visualización por más de unos segundos, y no tiene conocimiento técnico.
  3. Tienen que incluir además una *breve* descripción de cómo obtuvieron los datos que están presentando, que no entraría en el tweet.
